### ISSUE TITLE
Documentation on has_time parameter

### ISSUE THREAD

**Issue by jcrichard** [catboost/catboost#1076]
Problem:
The documentation is not that very clear on the parameter. If I have a dataset sorted by time and I put has_time=True does catboost do random permutations or not? 
There is apparently a way of adding a Datetime column but how do you do that with pool object?

**Comment by MarkusBerroth** [comment]
I am wondering about the same issue.

**Comment by RivermanPaul** [comment]
Would love some examples on how to use has_time with a regressor.  

I have a "year_sold" column (e.g. 2005) and a "month_sold" column (1-12) to represent when the sales data took place.  What do I need to do with these?

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
Dataset is permuted only if it contains categorical features and has_time=False.

Dataset w/o categorical features is never permuted because permutation does not change loss value.

Feel free to reopen, if you have further questions.

**Comment by RivermanPaul** [comment]
Thank you for you help!  I did have a couple more quick questions:

> Dataset w/o categorical features is never permuted because permutation does not change loss value.

Could you explain more about why permutation only changes the loss value if there are  categorical features?

If I understand you correctly, let’s say I do have some categorical features, but no timestamp column, and I set has_time = true.  Even though I have not included a time stamp column, the data will be used in order, not permuted?

In order to get the best results from CatBoost, would it be best to make a time stamp column and order my data by that column, or should I simply include the year/month as a numerical column, such as 2005.6 (should I perhaps try both)?

If my data is ordered by time stamp, should my validation data come strictly after my training data (in time), and should my testing data be the most recent of all?

**Comment by jcrichard** [comment]
I do not think the question has been answered fully. In the doc it also says that there is also permutation choosing the tree structure. Also, it is not clear whether we have to provide a datetime column or not. Could you be more specific?

**Comment by LyzhinIvan** [comment]
> Could you explain more about why permutation only changes the loss value if there are categorical features?

Because permutation is used only for transforming categorical features. https://catboost.ai/docs/concepts/algorithm-main-stages_cat-to-numberic.html#algorithm-main-stages_cat-to-numberic

> If I understand you correctly, let’s say I do have some categorical features, but no timestamp column, and I set has_time = true. Even though I have not included a time stamp column, the data will be used in order, not permuted?

Yes, with has_time=True and without Timestamp column Catboost will use initial dataset order. With has_time=True and with Timestamp column Catboost will use 1 permutation ordered by Timestamp column. Without has_time=True Catboost will generate 3 random permutations regardless of Timestamp column presence (if categorical features are present).

> In order to get the best results from CatBoost, would it be best to make a time stamp column and order my data by that column, or should I simply include the year/month as a numerical column, such as 2005.6 (should I perhaps try both)?

It depends on your task actually. But in general case I think the best approach will be to include Timestamp column and different parts of timestamp (year/month/day/hour/etc) as numeric features.

> If my data is ordered by time stamp, should my validation data come strictly after my training data (in time), and should my testing data be the most recent of all?

Catboost doesn't check it but usually it is required to avoid target leakage.

> I do not think the question has been answered fully. In the doc it also says that there is also permutation choosing the tree structure. Also, it is not clear whether we have to provide a datetime column or not. Could you be more specific?

Catboost uses 3 random permutations to transform categorical features to numerical. For each tree one of these permutations is used. About datetime column I answered above.

**Comment by antipisa** [comment]
> > Could you explain more about why permutation only changes the loss value if there are categorical features?
> 
> Because permutation is used only for transforming categorical features. https://catboost.ai/docs/concepts/algorithm-main-stages_cat-to-numberic.html#algorithm-main-stages_cat-to-numberic
> 
> > If I understand you correctly, let’s say I do have some categorical features, but no timestamp column, and I set has_time = true. Even though I have not included a time stamp column, the data will be used in order, not permuted?
> 
> Yes, with has_time=True and without Timestamp column Catboost will use initial dataset order. With has_time=True and with Timestamp column Catboost will use 1 permutation ordered by Timestamp column. Without has_time=True Catboost will generate 3 random permutations regardless of Timestamp column presence (if categorical features are present).
> 
> > In order to get the best results from CatBoost, would it be best to make a time stamp column and order my data by that column, or should I simply include the year/month as a numerical column, such as 2005.6 (should I perhaps try both)?
> 
> It depends on your task actually. But in general case I think the best approach will be to include Timestamp column and different parts of timestamp (year/month/day/hour/etc) as numeric features.
> 
> > If my data is ordered by time stamp, should my validation data come strictly after my training data (in time), and should my testing data be the most recent of all?
> 
> Catboost doesn't check it but usually it is required to avoid target leakage.
> 
> > I do not think the question has been answered fully. In the doc it also says that there is also permutation choosing the tree structure. Also, it is not clear whether we have to provide a datetime column or not. Could you be more specific?
> 
> Catboost uses 3 random permutations to transform categorical features to numerical. For each tree one of these permutations is used. About datetime column I answered above.

Does this happen with boosting_type="Plain" as well as "Ordered"?

**Comment by HananInfiniGrow** [comment]
what if I split my data to "train" "test" where "train" is in the past? why should time column matter than? so I don't understand what is the point behind `has_time`?
